dbms/src/Compression/CompressionCodecDoubleDelta.cpp

```diff
@@ -0,0 +1,320 @@
+            UInt8 i = 0;
+            for (; i < sizeof(data_sizes) - 1; ++i)
+            {
+                const auto next_bit = reader.readBit();
+                if (next_bit == 0)
+                    break;
+            }
+
+            const UInt8 sign = reader.readBit();
+            double_delta = static_cast<DeltaType>(reader.readBits(data_sizes[i]));
+            if (sign)
+            {
+                double_delta *= -1;
+            }
+        }
+        const T curr_value = static_cast<T>(prev_value + prev_delta + double_delta);
+        unalignedStore(dest, curr_value);
+        dest += sizeof(curr_value);
+
+        prev_delta = curr_value - prev_value;
+        prev_value = curr_value;
+    }
+}
+
+UInt8 getDataBytesSize(DataTypePtr column_type)
+{
+    UInt8 data_bytes_size = 1;
+    if (column_type && column_type->haveMaximumSizeOfValue())
+    {
+        size_t max_size = column_type->getSizeOfValueInMemory();
+        if (max_size == 1 || max_size == 2 || max_size == 4 || max_size == 8)
+            data_bytes_size = static_cast<UInt8>(max_size);
+    }
+    return data_bytes_size;
+}
+
+} // namespace
+
+
+CompressionCodecDoubleDelta::CompressionCodecDoubleDelta(UInt8 data_bytes_size_)
+    : data_bytes_size(data_bytes_size_)
+{
+}
+
+UInt8 CompressionCodecDoubleDelta::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::DoubleDelta);
+}
+
+String CompressionCodecDoubleDelta::getCodecDesc() const
+{
+    return "DoubleDelta";
+}
+
+UInt32 CompressionCodecDoubleDelta::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    const auto result = 2 // common header
+            + data_bytes_size // max bytes skipped if source is not properly aligned.
+            + getCompressedHeaderSize(data_bytes_size) // data-specific header
+            + getCompressedDataSize(data_bytes_size, uncompressed_size);
+
+    return result;
+}
+
+UInt32 CompressionCodecDoubleDelta::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    UInt8 bytes_to_skip = source_size % data_bytes_size;
+    dest[0] = data_bytes_size;
+    dest[1] = bytes_to_skip;
+    memcpy(&dest[2], source, bytes_to_skip);
+    size_t start_pos = 2 + bytes_to_skip;
+    UInt32 compressed_size = 0;
+    switch (data_bytes_size)
+    {
+    case 1:
+        compressed_size = compressDataForType<UInt8, Int16>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
+        break;
+    case 2:
+        compressed_size = compressDataForType<UInt16, Int32>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
+        break;
+    case 4:
+        compressed_size = compressDataForType<UInt32, Int64>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
+        break;
+    case 8:
+        compressed_size = compressDataForType<UInt64, Int64>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
```

Comment:
Is it enough to have Int64 for any combinations of UInt64? Do we throw out of range exception otherwise?

Reply:
I think that I have a test case just for that, could you please validate it? https://github.com/yandex/ClickHouse/pull/5600/files#r293377862